pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/groupby/DefaultGroupByExecutor.java

```diff
@@ -123,11 +130,22 @@ public void process(int[] docIdSet, int startIndex, int length) {
     Preconditions
         .checkState(_inited, "Method 'process' cannot be called before 'init' for class " + getClass().getName());
 
-    _singleValueBlockCache.initNewBlock(docIdSet, startIndex, length);
+    // We have already reached max number of keys, punt on further processing.
+    if (_groupKeyGenerator.getCurrentGroupKeyUpperBound() > _numGroupsLimit) {
```

Comment:
This might be ok to catch corner cases. How about continuing to count with the existing group keys already seen?